################################################################################
SHORT DESCRIPTION: 
################################################################################
QSB-023 (https://github.com/QubesOS/qubes-secpack/blob/master/QSBs/qsb-023-2015)
XSA-155 additional patches for the Xen network and block frontends from Qubes
OS Project.

################################################################################
LONG DESCRIPTION: 
################################################################################
Source: https://github.com/QubesOS/qubes-linux-kernel/tree/stable-3.18/patches.xen
Patch: xsa155-linux318-0010-xen-netfront-do-not-use-data-already-exposed-to-back.patch

Backend may freely modify anything on shared page, so use data which was
supposed to be written there, instead of reading it back from the shared
page.

################################################################################
PATCHES 
################################################################################
Index: linux-3.18.25/drivers/net/xen-netfront.c
===================================================================
--- linux-3.18.25.orig/drivers/net/xen-netfront.c	2015-12-22 15:06:54.122094546 +0100
+++ linux-3.18.25/drivers/net/xen-netfront.c	2015-12-22 15:07:15.328542483 +0100
@@ -457,6 +457,7 @@
 	int frags = skb_shinfo(skb)->nr_frags;
 	unsigned int offset = offset_in_page(data);
 	unsigned int len = skb_headlen(skb);
+	unsigned int size;
 	unsigned int id;
 	grant_ref_t ref;
 	int i;
@@ -464,10 +465,11 @@
 	/* While the header overlaps a page boundary (including being
 	   larger than a page), split it it into page-sized chunks. */
 	while (len > PAGE_SIZE - offset) {
-		tx->size = PAGE_SIZE - offset;
+		size = PAGE_SIZE - offset;
+		tx->size = size;
 		tx->flags |= XEN_NETTXF_more_data;
-		len -= tx->size;
-		data += tx->size;
+		len -= size;
+		data += size;
 		offset = 0;
 
 		id = get_id_from_freelist(&queue->tx_skb_freelist, queue->tx_skbs);
